File: src/cache.ts
import { createHash } from "node:crypto";
import { mkdir, readFile, rename, rm, writeFile } from "node:fs/promises";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import type {
  Account,
  Company,
  Instrument,
  Merchant,
  Tag,
  Transaction,
  User,
} from "./api.js";

export const CACHE_VERSION = 1;

export interface CacheData {
  version: number;
  /** Unix seconds when this snapshot was written. */
  savedAt: number;
  serverTimestamp: number;
  accounts: Account[];
  tags: Tag[];
  merchants: Merchant[];
  companies: Company[];
  instruments: Instrument[];
  transactions: Transaction[];
  users: User[];
}

/**
 * Stable, non-reversible id for a token. Used as the cache file name so
 * separate accounts never share a snapshot and the token itself is never
 * written to disk.
 */
export function tokenHash(token: string): string {
  return createHash("sha256").update(token).digest("hex").slice(0, 32);
}

export function cacheBaseDir(): string {
  const override = process.env.ZENMONEY_CACHE_DIR;
  if (override) return override;
  if (process.platform === "win32" && process.env.LOCALAPPDATA) {
    return join(process.env.LOCALAPPDATA, "zenmoney-mcp");
  }
  if (process.env.XDG_CACHE_HOME) {
    return join(process.env.XDG_CACHE_HOME, "zenmoney-mcp");
  }
  return join(homedir(), ".cache", "zenmoney-mcp");
}

/**
 * File-backed snapshot of the synced state, keyed by the hash of the API
 * token. Lets short-lived stdio processes reuse a previous full sync instead
 * of re-downloading everything on every launch.
 */
export class StateCache {
  readonly path: string;
  private writeChain: Promise<void> = Promise.resolve();

  constructor(token: string, dir: string = cacheBaseDir()) {
    this.path = join(dir, `${tokenHash(token)}.json`);
  }

  async load(): Promise<CacheData | null> {
    let raw: string;
    try {
      raw = await readFile(this.path, "utf8");
    } catch {
      return null;
    }

    try {
      const parsed = JSON.parse(raw) as CacheData;
      if (parsed?.version !== CACHE_VERSION) return null;
      if (typeof parsed.serverTimestamp !== "number") return null;
      return parsed;
    } catch {
      // Corrupted snapshot — treat it as a cache miss and let a full sync
      // overwrite it.
      return null;
    }
  }

  /**
   * Atomically replace the snapshot. Writes are chained so concurrent callers
   * cannot interleave into a half-written file.
   */
  async save(data: Omit<CacheData, "version" | "savedAt">): Promise<void> {
    const payload: CacheData = {
      ...data,
      version: CACHE_VERSION,
      savedAt: Math.floor(Date.now() / 1000),
    };

    this.writeChain = this.writeChain.then(() => this.writeSnapshot(payload));
    return this.writeChain;
  }

  private async writeSnapshot(payload: CacheData): Promise<void> {
    const dir = dirname(this.path);
    const tmp = `${this.path}.${process.pid}.tmp`;
    await mkdir(dir, { recursive: true, mode: 0o700 });
    await writeFile(tmp, JSON.stringify(payload), { mode: 0o600 });
    await rename(tmp, this.path);
  }

  async clear(): Promise<void> {
    await rm(this.path, { force: true });
  }
}

File: go/internal/zen/cache.go
package zen

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sync"
	"time"
)

// CacheVersion 2: reminders and their markers joined the snapshot. A version 1
// file has none of them, and an incremental sync from its timestamp would never
// report the ones that already exist, so old snapshots are dropped rather than
// read.
const CacheVersion = 2

// CacheData is the on-disk snapshot. Field names match the TypeScript server's
// file byte for byte so both implementations can share one cache directory.
type CacheData struct {
	Version int `json:"version"`
	// SavedAt is the unix second the snapshot was written.
	SavedAt         int64            `json:"savedAt"`
	ServerTimestamp int64            `json:"serverTimestamp"`
	Accounts        []Account        `json:"accounts"`
	Tags            []Tag            `json:"tags"`
	Merchants       []Merchant       `json:"merchants"`
	Companies       []Company        `json:"companies"`
	Instruments     []Instrument     `json:"instruments"`
	Transactions    []Transaction    `json:"transactions"`
	Users           []User           `json:"users"`
	Reminders       []Reminder       `json:"reminders"`
	ReminderMarkers []ReminderMarker `json:"reminderMarkers"`
}

// TokenHash is a stable, non-reversible id for a token. It names the cache file
// so separate accounts never share a snapshot and the token itself is never
// written to disk.
func TokenHash(token string) string {
	sum := sha256.Sum256([]byte(token))
	return hex.EncodeToString(sum[:])[:32]
}

// CacheBaseDir mirrors the TypeScript server's directory choice, including the
// ZENMONEY_CACHE_DIR override, so a snapshot written by either implementation
// is found by the other.
func CacheBaseDir() string {
	if override := os.Getenv("ZENMONEY_CACHE_DIR"); override != "" {
		return override
	}
	if runtime.GOOS == "windows" {
		if local := os.Getenv("LOCALAPPDATA"); local != "" {
			return filepath.Join(local, "zenmoney-mcp")
		}
	}
	if xdg := os.Getenv("XDG_CACHE_HOME"); xdg != "" {
		return filepath.Join(xdg, "zenmoney-mcp")
	}
	home, err := os.UserHomeDir()
	if err != nil {
		home = "."
	}
	return filepath.Join(home, ".cache", "zenmoney-mcp")
}

// Cache is a file-backed snapshot of the synced state, keyed by the hash of the
// API token. It lets short-lived processes reuse a previous full sync instead of
// re-downloading everything on every launch.
type Cache struct {
	Path string
	// Writes are serialized so concurrent callers cannot interleave into a
	// half-written file.
	mu sync.Mutex
}

func NewCache(token string) *Cache {
	return NewCacheIn(token, CacheBaseDir())
}

func NewCacheIn(token, dir string) *Cache {
	return &Cache{Path: filepath.Join(dir, TokenHash(token)+".json")}
}

// Load returns the snapshot, or nil on a cache miss, a version mismatch or a
// corrupted file — all of which a full sync overwrites anyway.
func (c *Cache) Load() *CacheData {
	raw, err := os.ReadFile(c.Path)
	if err != nil {
		return nil
	}
	var data CacheData
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil
	}
	if data.Version != CacheVersion {
		return nil
	}
	return &data
}

// Save atomically replaces the snapshot.
func (c *Cache) Save(data CacheData) error {
	data.Version = CacheVersion
	data.SavedAt = time.Now().Unix()

	c.mu.Lock()
	defer c.mu.Unlock()

	payload, err := json.Marshal(data)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(c.Path), 0o700); err != nil {
		return err
	}
	tmp := fmt.Sprintf("%s.%d.tmp", c.Path, os.Getpid())
	if err := os.WriteFile(tmp, payload, 0o600); err != nil {
		return err
	}
	if err := os.Rename(tmp, c.Path); err != nil {
		os.Remove(tmp)
		return err
	}
	return nil
}

func (c *Cache) Clear() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if err := os.Remove(c.Path); err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}
